### PR TITLE
service/block: add basic encoding validation in `BlockService`

### LINKED ISSUE #92: service/block: basic validation of DAH hash against raw header data root
Even without the possibility to gossip erasure coding fraud proofs, we should validate that the data root in the header matches the daheader somewhere here (ensure dah.Hash() == raw.Header.DataHash).

Core already has the validation logic (`ValidateBasic`) - this can just be re-used.

### PR DESCRIPTION
**Pull request by renaynay**
This PR relies on #87 being merged first.

Resolves #92.

### CODE REVIEW

File: service/block/event_test.go [hunk 1]
```diff
@@ -5,7 +5,7 @@ import (
 	"testing"
 
 	"github.com/celestiaorg/celestia-core/testutils"
-
+	"github.com/celestiaorg/celestia-node/service/header"
 	"github.com/stretchr/testify/require"
```
Comment (liamsi) [inline service/block/event_test.go:9]: Can someone please enable the linter s.t. we have a canonical order of packages?
```
std 

third party

this repo
```

File: service/block/fraud.go [hunk 2]
```diff
@@ -1,10 +1,19 @@
...
 // validateEncoding validates the erasure coding DataRoot in the raw block header against the
 // hash of the generated DataAvailabilityHeader, generating a fraud proof if there is a mismatch.
-// TODO @renaynay: this method is stubbed out for now.
-func (s *Service) validateEncoding(extendedBlock *ExtendedBlock, rawHeader core.Header) error {
+// TODO @renaynay: this method is stubbed out for now. It only performs a basic validation to check
+// whether the hash of the DAH equals the data root in the raw header.
+func validateEncoding(block *Block, rawHeader header.RawHeader) error {
+	if w, g := block.header.DAH.Hash(), rawHeader.DataHash; !bytes.Equal(w, g) {
+		return fmt.Errorf("wrong DataHash in RawHeader. Expected %X, got %X", w, g)
+	}
```
Comment (liamsi) [inline service/block/fraud.go:17]: 👍🏼

File: service/block/fraud_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"testing"
+
+	"github.com/celestiaorg/celestia-core/testutils"
+	"github.com/celestiaorg/celestia-node/service/header"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_validateEncoding_Successful(t *testing.T) {
```
Comment (liamsi) [inline service/block/fraud_test.go:11]: General remark: I'd prefer table-driven tests and also a few non-happy-path cases always.

**Review by liamsi — APPROVED** [review]
Looks good! 👏🏼

File: service/block/event_test.go (hunk 1 above)
Comment (renaynay) [reply service/block/event_test.go:9]: @Bidon15 ?

**Review by liamsi — APPROVED** [review]

File: service/block/fraud_test.go (hunk 3 above)
Comment (renaynay) [reply service/block/fraud_test.go:11]: Oh yes I prefer them too. I forgot to add the unhappy case.

File: service/block/event_test.go (hunk 1 above)
Comment (Wondertan) [reply service/block/event_test.go:9]: I agree, we need to have this how, checking this ourselves is annoying

File: service/block/fraud.go (hunk 2 above)
Comment (Wondertan) [reply service/block/fraud.go:17]: Header validation != encoding validation. I mean that this is the wrong place for the check or validateEncoding should be renamed to something like `validate` only

**Review by Wondertan — COMMENTED** [review]
It is not absolutely necessary to do this validation though, as we trust the Core. For the future case with p2p Fetcher, the protocol doing the fetch would do this check on the lower level as well. Also, if we do this check why don't we check other fields in the block validation process?  It just does not make sense to me not to trust the Core only for DataHash, but trust with other things.

cc @liamsi

**Comment by liamsi** [comment]
> It just does not make sense to me not to trust the Core only for DataHash, but trust with other things.

Yeah, I think we should ideally do basic validation on the block. But as we get header and block separentley, we should make sure that they match up (each field). 

> as we trust the Core.

Yeah, you are absolutely right about trusting core. I think we should still make sure that the stuff we get from core actually checks out. The check here is not wrong though.  I think we should open an issue for the checks but most likely we can just use whatever core does as a library with the difference that we download block and header separately here (but maybe there are functions we can still use in that case).

**Comment by renaynay** [comment]
>  think we should open an issue for the checks but most likely we can just use whatever core does as a library with the difference that we download block and header separately here (but maybe there are functions we can still use in that case).

@liamsi can you pls open this issue on our repo? 🙏🏻

**Review by Wondertan — APPROVED** [review]
Codewise looks good. Conceptually, my concern is still the same, we should either have validation for everything or for nothing, but we can see this as a first step towards validation for everything

**Pull request merged**